Credit provider integration Python API (Part 1 of 3)

wedaly opened the pull request:
This is the first of three pull requests for integration with credit providers.  This PR adds the Django models for tracking credit request and status, as well as the Python APIs for creating, retrieving, and updating requests.

* Add Python API functions for creating and updating requests for credit.
* Add Django models and migrations for tracking credit requests and status.

Future PRs: add HTTP end-points, including authorization checks; implement the "request credit" button.

@clintonb please review.

FYI: @aamir-khan @ahsan-ul-haq please take a look at how this code checks for credit eligibility and retrieves the final grade from the credit eligibility table.  If you have any questions, please let me know.

JIRA: [ECOM-1536](https://openedx.atlassian.net/browse/ECOM-1536)

clintonb commented on openedx/core/djangoapps/credit/api.py:
```diff
@@ -111,8 +136,212 @@ def get_credit_requirements(course_key, namespace=None):
+    Initiate a request for credit from a credit provider.
+
+    This will return the parameters that the user's browser will need to POST
+    to the credit provider.  It does NOT calculate the signature
```
Minor: add a `.`.

clintonb commented on openedx/core/djangoapps/credit/api.py:
```diff
@@ -111,8 +136,212 @@ def get_credit_requirements(course_key, namespace=None):
+    except CreditEligibility.DoesNotExist:
+        raise UserIsNotEligible
+    else:
+        credit_course = user_eligibility.course
```
Minor: Why not move this into the `try` block?

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -186,8 +196,141 @@ class CreditEligibility(TimeStampedModel):
+        try:
+            return self.statuses.latest().status  # pylint: disable=no-member
+        except CreditRequestStatus.DoesNotExist:
+            return "pending"
```
Make this a constant so that it can be shared elsewhere.

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -186,8 +196,141 @@ class CreditEligibility(TimeStampedModel):
+        unique_together = ('username', 'course', 'provider')
+
+
+class CreditRequestStatus(TimeStampedModel):
```
Minor: [django-simple-history](https://django-simple-history.readthedocs.org/en/latest/) would allow you to track all changes to `CreditRequest`, and remove the need for this model.

clintonb replied:
P.S. This would also eliminate the need for the raw query/subquery in `CreditRequest.credit_requests_for_user`.

wedaly replied:
Great suggestion!  Done.

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -18,6 +18,16 @@
+    Each provider is identified by unique ID (e.g., 'ASU').
+    """
+
+    provider_id = models.CharField(max_length=255, db_index=True, unique=True)
```
Minor: This name is too close to `id` which is already a field on the model. Consider changing to `code` or something that cannot be confused with the existing `id` field.

wedaly replied:
I hear the concern, but I'm inclined to keep it as `provider_id` to avoid introducing more conflicts with the other branches in-flight (we're all working off a design spec that calls it `provider_id`)

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -186,8 +196,141 @@ class CreditEligibility(TimeStampedModel):
+            (
+                'SELECT cr.uuid, cr.created, cc.course_key, cp.provider_id, cp.display_name, '
+                'IFNULL(('
+                '    SELECT status FROM credit_creditrequeststatus AS crs '
```
Use table names from the model's metadata (e.g. `CreditRequestStatus._meta.db_table`).

clintonb commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -1,29 +1,66 @@
+    def add_credit_course(self, enabled=True):
+        """Mark the course as a credit """
+        credit_course = CreditCourse.objects.create(course_key=self.course_key, enabled=enabled)
+        credit_course.save()
```
No need for the call to `save()`.

clintonb commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -1,29 +1,66 @@
+        credit_course.save()
+
+        # Associate a credit provider with the course.
+        credit_provider = CreditProvider(provider_id=self.PROVIDER_ID, display_name=self.PROVIDER_NAME)
```
Use `CreditProvider.objects.create()` to eliminate the call to `save()`.

clintonb commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -52,12 +89,9 @@ def setUp(self, **kwargs):
+            api.set_credit_requirements(self.course_key, requirements)
 
     def test_set_credit_requirements_invalid_course(self):
-        """Test that 'InvalidCreditCourse' exception is raise if we try to
```
Some documentation is needed to aid future developers working with these tests.

clintonb commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -67,16 +101,13 @@ def test_set_credit_requirements_invalid_course(self):
+            api.set_credit_requirements(self.course_key, requirements)
 
     def test_set_get_credit_requirements(self):
-        """Test that if same requirement is added multiple times
```
Some documentation is needed to aid future developers working with these tests.

wedaly replied:
Nose has this irritating habit of using the docstring as the test name in its output, which can make it difficult to find the method corresponding to a test failure.

I've added these docstrings back in as comments; I've also updated some of the new tests with comments describing what they're testing.

clintonb replied:
There is a plugin for that: https://pypi.python.org/pypi/nose-ignore-docstring.

clintonb commented on openedx/core/djangoapps/credit/tests/test_models.py:
```diff
@@ -83,3 +90,96 @@ def add_credit_course(self):
+        self.assertEqual(request.current_status(), "pending")
+
+        # Update the status
+        CreditRequestStatus.objects.create(request=request, status="approved")
```
A helper method on `CreditRequest` (e.g. `CreditRequest.set_status`) might prove useful for testing and in the future integration.

clintonb commented on openedx/core/djangoapps/credit/tests/test_models.py:
```diff
@@ -83,3 +90,96 @@ def add_credit_course(self):
+        self.credit_provider = CreditProvider.objects.create(provider_id=self.PROVIDER_ID)
+
+    def test_credit_request_current_status(self):
+        request = CreditRequest.objects.create(
```
Docstrings would be really helpful for future development and understanding these tests and their expectations.

clintonb commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -146,45 +173,220 @@ def test_requirements_to_disable(self):
+        self.user.profile.save()
+
+    def test_credit_request(self):
+        self._configure_credit()
```
This seems to be used in all the tests. It should be in `setUp()`. Hmm...not sure what to do about `test_user_is_not_eligible` aside from deleting all related objects and calling `_configure_credit()` with the appropriate status..

clintonb commented:
This is fine functionally, but I'd like to see more comments for the tests. :+1:

wedaly commented:
@clintonb Thanks for the review.  Rebased onto master and added an additional commit addressing your comments.

@feanil This PR now introduces a new dependency to edx-platform (django-simple-history) that is being used in the E-Commerce service.  It also includes a database migration, but none of the tables will be used in prod until this feature is enabled.

feanil commented:
@wedaly what will this history table be used for?  Will it be used in production?

wedaly commented:
@feanil It's an audit trail of requests to the credit provider and any responses we receive.

feanil commented:
Notes from talking to @wedaly offline.
- The history table will not be used by application code
- It is meant to be used as an audit trail and for offline reporting tools
- The history table should not grow very quickly because for each request in the credit table, there will be <= 1 response from the credit provider (approved / denied)
  - This would mean that the history table should grow linearly w.r.t the credit tables.

Given this info, the history table related changes make sense.

@wedaly  I also noticed that you are deleting two migrations and replacing it with one new one?  It looked like this is a consolidation?  When were the other two merged to master? It is possible they may have already been run on some of our systems.

wedaly commented:
@feanil Good catch -- looks like I made a mistake when rebasing this onto master.  I'll update the PR to avoid deleting migrations in master.

feanil commented:
@wedaly cool, migrations look good now.

feanil commented:
@wedaly noticed the django admin for this app is quite sparse: https://github.com/edx/edx-platform/blob/will/credit-provider-integration/openedx/core/djangoapps/credit/admin.py

Are there things that we might need to do operationally for this app from the admin console?  Will that be taken care of by some other user facing view?  What happens if we need to override a decision?

wedaly commented:
@feanil Almost all of the configuration for credit courses is going to be handled by the Course Admin tool that the architecture team is currently building, with the exception of setting up new credit providers.  I'm open to exposing more of this functionality through Django admin if we need to manually override anything, but I think that's outside the scope of this PR.

wedaly commented:
test this please

feanil commented:
:+1:   @wedaly sounds good, I just wanted to make sure there was a plan, sounds like there is one
